activemq-unit-tests/src/test/java/org/apache/activemq/jms2/ActiveMQJMS2ContextTest.java

```diff
@@ -0,0 +1,302 @@
+            }
+        }
+        
+        // Attempt to obtain a third context after all contexts have been closed
+        boolean caught = false;
+        try {
+            thirdJMSContext = secondJMSContext.createContext(Session.AUTO_ACKNOWLEDGE);
+            fail("JMSRuntimeException expected");
+        } catch (JMSRuntimeException e) {
+            caught = true;
+            assertEquals("Context already closed", e.getMessage());
+        }
+        assertTrue(caught);
+    }
+
+    @Test
+    public void testConnectionFactoryCreateContextBrowse() {
+        try(JMSContext jmsContext = activemqConnectionFactory.createContext()) {
+            assertNotNull(jmsContext);
+            jmsContext.start();
+            assertTrue(ActiveMQContext.class.isAssignableFrom(jmsContext.getClass()));
+            sendMessage(jmsContext, methodNameDestinationName);
+            browseMessage(jmsContext, methodNameDestinationName, "Test-" + methodNameDestinationName, true);
+        } catch (JMSException e) {
+            fail(e.getMessage());
+        }
+    }
+
+    @Test
+    public void testConnectionFactoryCreateContextBrowseAutoStart() {
+        try(JMSContext jmsContext = activemqConnectionFactory.createContext()) {
+            assertNotNull(jmsContext);
+            assertTrue(ActiveMQContext.class.isAssignableFrom(jmsContext.getClass()));
+            sendMessage(jmsContext, methodNameDestinationName);
+            browseMessage(jmsContext, methodNameDestinationName, "Test-" + methodNameDestinationName, true);
+        } catch (JMSException e) {
+            fail(e.getMessage());
+        }
+    }
+
+    @Test
+    public void testConnectionFactoryCreateContextBrowseAutoStartFalse() {
+        try(JMSContext jmsContext = activemqConnectionFactory.createContext()) {
+            assertNotNull(jmsContext);
+            jmsContext.setAutoStart(false);
+            assertTrue(ActiveMQContext.class.isAssignableFrom(jmsContext.getClass()));
+            sendMessage(jmsContext, methodNameDestinationName);
+            browseMessage(jmsContext, methodNameDestinationName, "Test-" + methodNameDestinationName, false);
+        } catch (JMSException e) {
+            fail(e.getMessage());
+        }
+    }
+
+    @Test
+    public void testConnectionFactoryCreateContextBrowseAutoStartFalseStartDelayed() {
+        try(JMSContext jmsContext = activemqConnectionFactory.createContext()) {
+            assertNotNull(jmsContext);
+            jmsContext.setAutoStart(false);
+            assertTrue(ActiveMQContext.class.isAssignableFrom(jmsContext.getClass()));
+            sendMessage(jmsContext, methodNameDestinationName);
+            jmsContext.start();
+            browseMessage(jmsContext, methodNameDestinationName, "Test-" + methodNameDestinationName, true);
+        } catch (JMSException e) {
+            fail(e.getMessage());
+        }
+    }
+
+    @Test
+    public void testDisableMessageID() {
```

Comment:
I don't see much in the way of JMSProducer testing, namely the tests that verify that a JMSProducer which is essentially a pooled anonymous producer which can have different configurations applied to it's different users simultaneously.  Tests that show a send from one with a particular configuration such as disable timestamps etc do not affect a different wrapped instance sending with some alternate configuration.

Reply:
This is done, unit tests have been added